flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/RecoveredChannelStateHandler.java

```diff
@@ -85,19 +84,15 @@
     @Override
     public void recover(InputChannelInfo channelInfo, int oldSubtaskIndex, Buffer buffer)
             throws IOException {
-        try {
-            if (buffer.readableBytes() > 0) {
-                for (final RecoveredInputChannel channel : getMappedChannels(channelInfo)) {
-                    channel.onRecoveredStateBuffer(
-                            EventSerializer.toBuffer(
-                                    new SubtaskConnectionDescriptor(
-                                            oldSubtaskIndex, channelInfo.getInputChannelIdx()),
-                                    false));
-                    channel.onRecoveredStateBuffer(buffer.retainBuffer());
-                }
+        if (buffer.readableBytes() > 0) {
+            for (final RecoveredInputChannel channel : getMappedChannels(channelInfo)) {
+                channel.onRecoveredStateBuffer(
+                        EventSerializer.toBuffer(
+                                new SubtaskConnectionDescriptor(
+                                        oldSubtaskIndex, channelInfo.getInputChannelIdx()),
+                                false));
+                channel.onRecoveredStateBuffer(buffer.retainBuffer());
             }
-        } finally {
-            buffer.recycleBuffer();
         }
     }
```

Comment:
1. I don't like that we are mixing different styles of contracts here. `RecoveredChannelStateHandler#recover` receives a `Buffer` that it doesn't own - caller is responsible for the clean up. While a couple lines below its copying buffer and passing the ownership of the copy to the `channel` - callee is responsible for the clean up. Why not stick with something that's consistent, and just have a contract that we are passing the ownership of the buffer to the callee? 
2. It's not documented anywhere.